File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        int[] currow = new int[obstacleGrid[0].length + 1];
        int[] dummy  = new int[obstacleGrid[0].length + 1];
        currow[n] = 1;
        for(int i = m - 1; i > -1; i--){
            for(int j = n-1; j > -1; j--){
                if(obstacleGrid[i][j] == 1){
                    currow[j] = 0;
                }
                else{
                    currow[j] = currow[j + 1] + dummy[j];
                }
            }
            dummy = currow;
            currow = new int[n + 1];
        }
        return dummy[0];
    }
}

File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        n = len(obstacleGrid[0])
        m = len(obstacleGrid)
        currow = [0] * (n + 1)
        dummy = [0] * n
        currow[-1] = 1
        for i in range(m - 1,-1, -1):
            for j in range(n-1, -1, -1):
                if obstacleGrid[i][j] == 1:
                    currow[j] = 0
                else:
                    currow[j] = currow[j + 1] + dummy[j]
            dummy = currow
            currow = [0] * (n + 1)
        return dummy[0]
